include/emp/web/Color.hpp

```diff
@@ -0,0 +1,366 @@
+
+        float alpha = 1.0f;
+        if (fname == "rgba" || fname == "rgb") {
+            return ParseRGB(str, css_str, fname, params, alpha);
+
+        } else if (fname == "hsla" || fname == "hsl") {
+            return ParseHSL(str, css_str, fname, params, alpha);
+        }
+        return *this;
+    }
+
+    emp_assert( false, css_str );
+    return *this;
+  }
+
+  bool Color::DetectABC(const std::string str) {
+      if (str.length() && str.front() == '#')
+          return true;
+      return false;
+  }
+  Color & Color::ParseABC(const std::string str, const std::string& css_str){
+      if (str.length() == 4) {
+            int64_t iv = color_impl::parseInt(str.substr(1), 16);  // TODO(deanm): Stricter parsing.
+            if (!(iv >= 0 && iv <= 0xfff)) {
+                emp_assert( false, css_str );
+                return *this;
+            } else {
+                *this = {
+                    static_cast<uint8_t>(((iv & 0xf00) >> 4) | ((iv & 0xf00) >> 8)),
+                    static_cast<uint8_t>((iv & 0xf0) | ((iv & 0xf0) >> 4)),
+                    static_cast<uint8_t>((iv & 0xf) | ((iv & 0xf) << 4)),
+                    1
+                };
+                return *this;
+            }
+        } else if (str.length() == 7) {
+            int64_t iv = color_impl::parseInt(str.substr(1), 16);  // TODO(deanm): Stricter parsing.
+            if (!(iv >= 0 && iv <= 0xffffff)) {
+              emp_assert( false, css_str );
+              return *this;  // Covers NaN.
+            } else {
+                *this = {
+                    static_cast<uint8_t>((iv & 0xff0000) >> 16),
+                    static_cast<uint8_t>((iv & 0xff00) >> 8),
+                    static_cast<uint8_t>(iv & 0xff),
+                    1
+                };
+                return *this;
+            }
+        }
+
+        emp_assert( false, css_str );
+        return *this;
+  }
+  
+  Color & Color::ParseRGB(const std::string str, const std::string& css_str, const std::string fname,
+        const std::vector<std::string> params, float alpha){
+      if (fname == "rgba") {
+                if (params.size() != 4) {
+                    emp_assert( false, css_str );
+                    return *this;
+                }
+                alpha = color_impl::parse_css_float(params.back());
+            } else {
+                if (params.size() != 3) {
+                    emp_assert( false, css_str );
+                    return *this;
+                }
+            }
+
+            *this = {
+                color_impl::parse_css_int(params[0]),
+                color_impl::parse_css_int(params[1]),
+                color_impl::parse_css_int(params[2]),
+                alpha
+            };
+            return *this;
+  }
+
+  Color& Color::ParseHSL(const std::string str, const std::string& css_str, const std::string fname,
+        const std::vector<std::string> params, float alpha){
+            if (fname == "hsla") {
```

Comment:
same comment --- instead of passing fname in, let's write two functions one for hsl and one for hsla

Reply:
still need to fix this